File: javascript/easy/implementStackUsingQueues.js
/**
 * Initialize your data structure here.
 */
var MyStack = function() {
    this.queue = [];
    this.queue2 = [];
    this.push_queue = this.queue;
    this.pop_queue = this.queue2;
};

/**
 * Push element x onto stack. 
 * @param {number} x
 * @return {void}
 */
MyStack.prototype.push = function(x) {
    this.push_queue.push(x);
};

/**
 * Removes the element on top of the stack and returns that element.
 * @return {number}
 */
MyStack.prototype.pop = function() {
    while(this.push_queue.length > 1) {
        this.pop_queue.push(this.push_queue.shift());
    }
    let result = this.push_queue.shift();
    this.swap();
    return result;
    return 0;
};

/**
 * Get the top element.
 * @return {number}
 */
MyStack.prototype.top = function() {
    while(this.push_queue.length > 1) {
        this.pop_queue.push(this.push_queue.shift());
    }
    let result = this.push_queue.shift();
    this.pop_queue.push(result);
    this.swap();
    return result;
};

/**
 * Returns whether the stack is empty.
 * @return {boolean}
 */
MyStack.prototype.empty = function() {
    return (this.push_queue.length == 0) && (this.pop_queue.length == 0);
};

MyStack.prototype.swap = function() {
    let temp = this.push_queue;
    this.push_queue = this.pop_queue;
    this.pop_queue = temp;
};

/** 
 * Your MyStack object will be instantiated and called as such:
 * var obj = Object.create(MyStack).createNew()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: ruby/easy/implement_stack_using_queues.rb
class MyStack

=begin
    Initialize your data structure here.
=end
    def initialize()
        @queue = Array.new
        @queue2 = Array.new
        @push_queue = @queue
        @pop_queue = @queue2
    end


=begin
    Push element x onto stack.
    :type x: Integer
    :rtype: Void
=end
    def push(x)
        @push_queue.push(x)
        pipe()
        swap()
    end


=begin
    Removes the element on top of the stack and returns that element.
    :rtype: Integer
=end
    def pop()
        res = @pop_queue.shift()
        return res
    end


=begin
    Get the top element.
    :rtype: Integer
=end
    def top()
        return @pop_queue.first
    end


=begin
    Returns whether the stack is empty.
    :rtype: Boolean
=end
    def empty()
        @push_queue.empty? and @pop_queue.empty?
    end

    private
    def pipe
        while !@pop_queue.empty?
            @push_queue.push(@pop_queue.shift())
        end
    end

    def swap
        temp = @push_queue
        @push_queue = @pop_queue
        @pop_queue = temp
    end
end

# Your MyStack object will be instantiated and called as such:
# obj = MyStack.new()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()